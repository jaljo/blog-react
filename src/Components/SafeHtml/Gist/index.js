import Gist from './Gist'
import { extractGistIdFromUrl } from './../../../Utils'
import { connect } from 'react-redux'
import { componentDidMount } from 'react-functional-lifecycle'
import { renderGist } from './../../../Redux/State/gists'
import { compose } from 'ramda'

// mapStateToProps :: (State, Props) -> Props
const mapStateToProps = (state, ownProps) => ({
  id: extractGistIdFromUrl(ownProps.url),
})

// mapDispatchToProps :: (Action * -> State) -> Props
const mapDispatchToProps = dispatch => ({
  renderGist: compose(dispatch, renderGist),
})

// didMount :: Props -> Action
const didMount = ({ id, renderGist }) => renderGist(id)

// lifecycles :: React.Component -> React.Component
const lifecycles = componentDidMount(didMount)(Gist)

// Gist :: Props -> React.Component
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(lifecycles)
