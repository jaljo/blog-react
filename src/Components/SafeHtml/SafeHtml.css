span.code {
  font-family: 'Courier New', Courier, monospace;
  background-color: #f7f7f7;
  border-radius: 3px;
  border: 1px solid #ddd;
  color: #e01e5a;
  font-size: 0.9rem;
  padding: 2px 3px;
}

blockquote {
  margin: 2.5rem 0;
  font-size: 1.5rem;
  font-style: italic;
  padding-left: 1rem;
  border-left: 3px solid #ddd;
  border-radius: 3px;
}

ul {
  list-style: square;
}

figure[data-is="image-with-caption"] {
  margin: 2.5rem 0;
}

figure[data-is="image-with-caption"] img {
  display: block;
  margin: auto;
  margin-bottom: 1rem;
}

figure[data-is="image-with-caption"] figcaption {
  font-size: 1rem;
  font-style: italic;
  text-align: center;
  width: 70%;
  margin: auto;
}
