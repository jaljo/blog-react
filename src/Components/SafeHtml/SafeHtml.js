import React from 'react'
import renderComponent from './SafeHtml'
import CreateGist from './Gist'
import ExternalLink from './../Router/Link/External'
import {
  cond,
  equals,
  o,
  prop,
} from 'ramda'
import {
  BLOCKQUOTE,
  BOLD,
  GIT_GIST,
  INLINE_CODE,
  ITALIC,
  LINK,
  LIST,
  LIST_ITEM,
  PARAGRAPH,
  TEXT,
  FIGURE,
} from './../../HtmlParser/HtmlParser'
import './SafeHtml.css'

// Text :: Component -> React.Component
export const Text = prop('content')

// Paragraph :: Component -> React.Component
export const Paragraph = ({ children }, idx) =>
  <p key={idx}>
    {children.map(renderComponent)}
  </p>

// Bold :: Component -> React.Component
export const Bold = ({ children }, idx) =>
  <b key={idx}>
    {children.map(renderComponent)}
  </b>

// Italic :: Component -> React.Component
export const Italic = ({ children }, idx) =>
  <i key={idx}>
    {children.map(renderComponent)}
  </i>

// Link :: Component -> React.Component
export const Link = ({ children, href }, idx) =>
  <ExternalLink to={href} key={idx}>
    {children.map(renderComponent)}
  </ExternalLink>

// Blockquote :: Component -> React.Component
export const Blockquote = ({ children }, idx) =>
  <blockquote key={idx} >
    {children.map(renderComponent)}
  </blockquote>

// Gist :: Component -> React.Component
export const Gist = (component, idx) =>
  <CreateGist
    key={idx}
    url={component['data-src']}
  />

// InlineCode :: Component -> React.Component
export const InlineCode = ({ children }, idx) =>
  <span key={idx} className="code">
    {children.map(renderComponent)}
  </span>

// List :: Component -> React.Component
export const List = ({ children }, idx) =>
  <ul key={idx}>
    {children.map(renderComponent)}
  </ul>

// ListItem :: Component -> React.Component
export const ListItem = ({ children }, idx) =>
  <li key={idx}>
    {children.map(renderComponent)}
  </li>

// Figure :: Component -> React.Component
export const Figure = ({ image, caption }, key) =>
  <figure key={key} data-is="image-with-caption">
    <img src={image} alt="" />
    <figcaption>
      {caption}
    </figcaption>
  </figure>

// is :: String -> Component -> Boolean
export const is = type => o(equals(type), prop('type'))

// SafeHtml :: Component -> React.Component
export default cond([
  [is(BLOCKQUOTE), Blockquote],
  [is(BOLD), Bold],
  [is(FIGURE), Figure],
  [is(GIT_GIST), Gist],
  [is(INLINE_CODE), InlineCode],
  [is(ITALIC), Italic],
  [is(LINK), Link],
  [is(LIST), List],
  [is(LIST_ITEM), ListItem],
  [is(PARAGRAPH), Paragraph],
  [is(TEXT), Text],
])
