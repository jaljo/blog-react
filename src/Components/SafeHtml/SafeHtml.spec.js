import * as SafeHtml from './SafeHtml'
import renderer from 'react-test-renderer';

describe('Components :: SafeHtml', () => {
  it('renders a paragraph component', () => {
    const tree = renderer.create(
      SafeHtml.Paragraph({ children: [
        { type: '@type/TEXT', content: 'hi there :)' }
      ]})
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders bold componet', () => {
    const tree = renderer.create(
      SafeHtml.Bold({ children: [
        { type: '@type/TEXT', content: 'a bold text' }
      ]})
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders a text component', () => {
    const tree = renderer.create(
      SafeHtml.Text({ content: 'a simple text' })
    )

    expect(tree).toMatchSnapshot()
  })

  it('identifies a component type', () => {
    const componentMock = { type: 'test' }

    expect(SafeHtml.is('test')(componentMock)).toBeTruthy()
    expect(SafeHtml.is('noop')(componentMock)).toBeFalsy()
  })

  it('renders a component tree to HTML', () => {
    const componentTreeMock = {
      type: '@type/PARAGRAPH',
      children: [
        {
          type: '@type/BOLD',
          children: [
            {
              type: '@type/TEXT',
              content: 'a test text'
            },
            {
              type: '@type/TEXT',
              content: 'another text'
            }
          ]
        },
      ]
    }

    const tree = renderer.create(
      SafeHtml.default(componentTreeMock)
    )

    expect(tree).toMatchSnapshot()
  })
})
