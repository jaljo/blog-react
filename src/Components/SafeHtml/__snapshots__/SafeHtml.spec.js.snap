// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components :: SafeHtml renders a blockquote component 1`] = `
<blockquote>
  a quoted text
</blockquote>
`;

exports[`Components :: SafeHtml renders a component tree to HTML 1`] = `
<p>
  <b>
    a test text
    another text
  </b>
  <a
    href="http://jlanglois.fr"
    rel="noopener noreferrer"
    target="_blank"
  >
    a link text child
  </a>
  <i>
    an italic text
  </i>
</p>
`;

exports[`Components :: SafeHtml renders a figure component 1`] = `
<figure
  data-is="image-with-caption"
>
  <img
    alt=""
    src="https://www.an-image.com/test.jpg"
  />
  <figcaption>
    this is the caption of the image
  </figcaption>
</figure>
`;

exports[`Components :: SafeHtml renders a gist component 1`] = `
<iframe
  data-is="gist"
  id="86954a679fcaed5e960f687ead25c4aa"
  title="86954a679fcaed5e960f687ead25c4aa"
/>
`;

exports[`Components :: SafeHtml renders a link component 1`] = `
<a
  href="test link"
  rel="noopener noreferrer"
  target="_blank"
>
  a link text
</a>
`;

exports[`Components :: SafeHtml renders a list component with items 1`] = `
<ul>
  <li>
    pin one
  </li>
  <li>
    pin two
  </li>
  <li>
    pin three
  </li>
</ul>
`;

exports[`Components :: SafeHtml renders a nested list with formatting components 1`] = `
<ul>
  <li>
    pin one
    <i>
      with some italic text
    </i>
    in it.
  </li>
</ul>
`;

exports[`Components :: SafeHtml renders a paragraph component 1`] = `
<p>
  hi there :)
</p>
`;

exports[`Components :: SafeHtml renders a text component 1`] = `"a simple text"`;

exports[`Components :: SafeHtml renders an inline code component 1`] = `
<span
  className="code"
>
  a piece of code
</span>
`;

exports[`Components :: SafeHtml renders an italic component 1`] = `
<i>
  an italic text
</i>
`;

exports[`Components :: SafeHtml renders bold component 1`] = `
<b>
  a bold text
</b>
`;
